File: typescript/questions/maxSlidingWindow.ts
function maxSlidingWindow(nums: number[], k: number): number[] {
  let nl = nums.length;
  let ans: number[] = new Array(nl - k + 1);
  let dq: number[] = [];
  for (let p = 0; p < nl; ++p) {
    while (dq.length && nums[p] > nums[dq[dq.length - 1]]) {
      dq.pop();
    }
    dq.push(p);
    let l = p - k + 1;
    if (dq[0] < l) {
      dq.shift();
    }
    if (l >= 0) {
      ans[l] = nums[dq[0]];
    }
  }
  return ans;
}

File: racket/questions/max_sliding_window.rs
use std::collections::VecDeque;

impl Solution {
    pub fn max_sliding_window(nums: Vec<i32>, k: i32) -> Vec<i32> {
        let k = k as usize;
        let nl = nums.len();
        let mut ans: Vec<i32> = Vec::with_capacity(nl - k + 1);
        let mut dq: VecDeque<usize> = VecDeque::new();
        for r in 0..nl {
            while let Some(&b) = dq.back() {
                if nums[r] > nums[b] {
                    dq.pop_back();
                } else {
                    break;
                }
            }
            dq.push_back(r);
            if dq[0] + k <= r {
                dq.pop_front();
            }
            if r >= k - 1 {
                ans.push(nums[dq[0]]);
            }
        }
        ans
    }
}
